`download-authz-both-sides` c4: a 403 attachment list is rendered as "No attachments yet"

## Symptom

A member who is denied the parent record opens a record that holds 2095+ attachments. The attachments panel fires `GET /api/v1/data/sys_attachment?...filter=parent_object=…`, the server returns **403**, and the panel renders its empty state — **"No attachments yet. Upload a file to get started."** — and still advertises an **Upload** affordance the server would refuse. An operator cannot tell "nothing here" from "not yours".

## Root cause

Not located by the run (objectui internals). The panel swallows a 40x list response into its empty state instead of mapping it to a distinct "denied" state. Nothing leaks — the objectui#2532 failure mode (raw error dump / leaked row) is **absent**, and no status code is shown; the gap is purely surfacing: a 40x is folded into the *empty* state rather than a *denied* state.

## Reproduction

As a member denied READ on the parent, open a record that in fact holds attachments. The list request returns 403 and the panel shows the empty state.

- Expected: denial copy (a "you don't have access to these attachments" state), and no Upload affordance.
- Actual: "No attachments yet. Upload a file to get started." + an Upload affordance, on a record that holds 2095+ attachments.

## Source

Extracted from the QA run objectstack-ai/objectstack#7635 (framework 92f26f75, console 6314e87f).

<!-- os-dev-report -->
```json
{
  "issue": 4269,
  "status": "done",
  "branch": "claude/issue-4269-attachments-denied-state",
  "pr": "https://github.com/objectstack-ai/objectui/pull/4685",
  "premise_still_valid": true,
  "summary": "Re-derived all three premises on current main (b8818171f) before building. Premise A holds and is now measured precisely: a 403 on the sys_attachment list read THROWS — ObjectStackAdapter.find() degrades only a non-authz 404 to {data: [], total: 0} and rethrows the rest (objectui#4408's carve-out) — so RecordAttachmentsPanel's bare `catch { setRows([]) }` was where 'denied' and 'empty' merged. Premise B: the Upload affordance was keyed on NOTHING (only `disabled={uploading}`), so withdrawing it under a denied list disturbs no existing permission-based visibility. Premise C: a house pattern exists and was mirrored rather than reinvented — `isPermissionError` from @object-ui/react (403 / PERMISSION_DENIED / FORBIDDEN / RLS), the same predicate ObjectKanban, ObjectCalendar and the form renderer branch on, plus ObjectDataPage's Lock-icon denial framing. The panel now renders a distinct denied state ('You don't have access to these attachments.', new detail.attachmentsAccessDenied key in en + all ten locale packs) with no Upload affordance, checked before the empty state; the empty state is reserved for a genuine 200-with-zero-rows. #2532's leak mode stays absent — the denied state renders the i18n sentence and nothing from the error, and setError is deliberately not called on that path. Non-authz failures keep today's handling, pinned by a test labelled as a pin, with the remaining swallow filed as #4684.",
  "tests": "All gates run AFTER the final commit, at HEAD 6e3013925 (working tree clean). (1) Panel tests — `pnpm exec vitest run packages/app-shell/src/views/__tests__/RecordAttachmentsPanel.test.tsx` => 'Test Files 1 passed (1) / Tests 14 passed (14)' (8 pre-existing + 6 new). (2) Full i18n suite (all ten packs touched) — `pnpm exec vitest run packages/i18n/` => '45 passed (45) / 804 passed (804)', includes all-locales-key-parity + en-zh-key-parity. (3) Build closure first: `pnpm --workspace-concurrency=2 --filter '@object-ui/app-shell^...' build` => exit 0. (4) `pnpm --filter @object-ui/app-shell --filter @object-ui/i18n type-check` => both 'Done' (NOTE: the script is `type-check`, not `typecheck`; the latter returned ERR_PNPM_RECURSIVE_RUN_NO_SCRIPT — caught, not silently passed). (5) `eslint --quiet` over all 12 changed source files => exit 0. (6) `pnpm run check:i18n-keys` => every in-scope call-site key resolves against the en pack, every literal inline defaultValue matches its en value. (7) `pnpm run check:i18n-drift` => '1 key(s) added, 0 en value(s) changed'. (8) `pnpm run check:control-bytes` => OK, 4209 tracked text files. (9) `node scripts/check-changeset-presence.mjs` => '12 source file(s) of 2 released package(s) changed, and this change declares 1 changeset(s)'. REVERSE VERIFICATION, direction predicted before running: reverting ONLY the component (tests kept) should turn exactly the 4 tests requiring the denied state red, while the 200-empty test and the non-authz PIN stay green (a pin asserts unchanged behaviour, so it must pass on both trees). Observed exactly that: 'Tests 4 failed | 10 passed (14)', and the failure dumps reproduce the reported symptom verbatim — on a 403 the reverted component rendered 'Upload' + 'No attachments yet. Upload a file to get started.' Restored byte-identically (`git diff HEAD` empty) and re-ran green at 14/14. Note: a repo guard (objectui#3288) rejected the first `pnpm --filter ... test -- --run path` form because vitest silently drops args after `--`; re-run in the prescribed repo-root form. No ablation in this change, so nothing is claimed about dist/.",
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #4684: RecordAttachmentsPanel still renders 'No attachments yet' for a network/5xx/401 failure it never loaded — same defect class one status over, deliberately left by this card's ruling; needs the loaded/unknown status vocabulary HomeActionCenter (#4235), the app list (#4300) and ListView.classifyLoadError already carry. Unassigned, unlabeled for PM triage; the pin test in this PR is the one to rewrite when it is picked up."
  ]
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/attachments-denied-vs-empty.md
+++ b/.changeset/attachments-denied-vs-empty.md
@@ -1,0 +1,21 @@
+---
+'@object-ui/app-shell': patch
+'@object-ui/i18n': patch
+---
+
+RecordAttachmentsPanel distinguishes a DENIED attachment list from an empty one.
+
+A `sys_attachment` list read refused for authorization reasons (HTTP 403 /
+`PERMISSION_DENIED` / `FORBIDDEN` / a row-level-security denial) was swallowed
+into the panel's empty state, so a member denied the parent record was told
+"No attachments yet. Upload a file to get started." about a record holding
+2095+ attachments — and was offered an Upload the server would refuse.
+
+The panel now classifies that refusal with the same `isPermissionError`
+predicate the kanban, calendar and form surfaces branch on, renders a distinct
+denied state ("You don't have access to these attachments.", new
+`detail.attachmentsAccessDenied` key in all ten locale packs), and withdraws
+the Upload affordance. The denied state renders the translated sentence and
+nothing sourced from the error — no status code, no server text, no row count.
+The empty state is now reserved for a genuine 200-with-zero-rows; non-authz
+failures keep their pre-existing handling.

--- a/packages/app-shell/src/views/RecordAttachmentsPanel.tsx
+++ b/packages/app-shell/src/views/RecordAttachmentsPanel.tsx
@@ -8,10 +8,10 @@
 
 import * as React from 'react';
 import { cn, Button } from '@object-ui/components';
-import { Paperclip, Upload, Trash2, Download, Loader2 } from 'lucide-react';
+import { Paperclip, Upload, Trash2, Download, Loader2, Lock } from 'lucide-react';
 import { createObjectStackUploadAdapter } from '@object-ui/providers';
 import { createAuthenticatedFetch } from '@object-ui/auth';
-import { useObjectTranslation } from '@object-ui/react';
+import { useObjectTranslation, isPermissionError } from '@object-ui/react';
 
 /**
  * RecordAttachmentsPanel — generic record Attachments surface (#2727,
@@ -69,6 +69,17 @@ export const RecordAttachmentsPanel: React.FC<RecordAttachmentsPanelProps> = ({
   const [loading, setLoading] = React.useState(false);
   const [uploading, setUploading] = React.useState(false);
   const [error, setError] = React.useState<string | null>(null);
+  /**
+   * The list read was refused for AUTHORIZATION reasons (#4269).
+   *
+   * Kept separate from `rows.length === 0` because the two say opposite
+   * things and only one of them is an assertion the panel is entitled to
+   * make. "No attachments yet" claims the record HOLDS nothing; a 403 says
+   * only that this caller may not look. Folding the second into the first
+   * told a denied member that a record with 2095+ attachments was empty —
+   * and offered an Upload the server would refuse.
+   */
+  const [listDenied, setListDenied] = React.useState(false);
   const inputRef = React.useRef<HTMLInputElement | null>(null);
 
   // Same base-URL convention as RecordDetailView's raw API fetches: the
@@ -131,10 +142,28 @@ export const RecordAttachmentsPanel: React.FC<RecordAttachmentsPanelProps> = ({
       });
       const items: AttachmentRow[] = Array.isArray(res) ? res : res?.data ?? [];
       setRows(items);
-    } catch {
-      // A 404 (table not provisioned on older stacks) is tolerated silently;
-      // the panel just stays empty.
+      setListDenied(false);
+    } catch (err) {
       setRows([]);
+      // An authorization refusal is NOT an empty record (#4269). The house
+      // predicate is the same one the kanban/calendar/form surfaces branch
+      // on — HTTP 403, `PERMISSION_DENIED`/`FORBIDDEN`, or an RLS denial.
+      // The adapter throws it: `find()` degrades only a non-authz 404 to
+      // `{ data: [], total: 0 }` (data-objectstack, objectui#4408), so a 403
+      // reaches this catch as a decorated throw.
+      //
+      // Nothing from the error is rendered — the denied state below shows the
+      // i18n sentence and nothing else. objectui#2532's failure mode (raw
+      // dump / status code / leaked row) must stay absent, and `setError` is
+      // deliberately NOT called here.
+      setListDenied(isPermissionError(err));
+      // Everything else keeps the pre-existing behaviour: a 404 (table not
+      // provisioned on older stacks) and any network/5xx failure are tolerated
+      // silently and the panel stays empty. That swallow is the SAME defect
+      // class one status over — an unreachable server also renders "No
+      // attachments yet" — but the honest unknown-vs-empty split is a separate
+      // change (filed, not fixed here) and this line pins today's behaviour
+      // rather than quietly widening the fix.
     } finally {
       setLoading(false);
     }
@@ -253,28 +282,39 @@ export const RecordAttachmentsPanel: React.FC<RecordAttachmentsPanelProps> = ({
             <span className="text-xs text-muted-foreground">({rows.length})</span>
           )}
         </div>
-        <div className="flex items-center gap-2">
-          <input
-            ref={inputRef}
-            type="file"
-            multiple
-            className="hidden"
-            onChange={(e) => void handleFiles(e.target.files)}
-          />
-          <Button
-            variant="outline"
-            size="sm"
-            disabled={uploading}
-            onClick={() => inputRef.current?.click()}
-          >
-            {uploading ? (
-              <Loader2 className="h-4 w-4 mr-1 animate-spin" />
-            ) : (
-              <Upload className="h-4 w-4 mr-1" />
-            )}
-            {t('detail.uploadAttachment', { defaultValue: 'Upload' })}
-          </Button>
-        </div>
+        {/*
+          No Upload affordance under a denied list (#4269). The button was
+          previously unconditional — its only gate was `uploading` — so a
+          member the server had just refused was still invited to upload into
+          a record whose parent it cannot read, a click the `beforeInsert`
+          gate answers with 403 ATTACHMENT_PARENT_ACCESS. Hiding it here
+          changes nothing for every other caller: this is the sole condition
+          added to a control that had none.
+        */}
+        {!listDenied && (
+          <div className="flex items-center gap-2">
+            <input
+              ref={inputRef}
+              type="file"
+              multiple
+              className="hidden"
+              onChange={(e) => void handleFiles(e.target.files)}
+            />
+            <Button
+              variant="outline"
+              size="sm"
+              disabled={uploading}
+              onClick={() => inputRef.current?.click()}
+            >
+              {uploading ? (
+                <Loader2 className="h-4 w-4 mr-1 animate-spin" />
+              ) : (
+                <Upload className="h-4 w-4 mr-1" />
+              )}
+              {t('detail.uploadAttachment', { defaultValue: 'Upload' })}
+            </Button>
+          </div>
+        )}
       </div>
 
       {error && (
@@ -287,6 +327,18 @@ export const RecordAttachmentsPanel: React.FC<RecordAttachmentsPanelProps> = ({
         <div className="px-4 py-6 text-sm text-muted-foreground flex items-center gap-2">
           <Loader2 className="h-4 w-4 animate-spin" />
           {t('detail.loadingAttachments', { defaultValue: 'Loading attachments…' })}
+        </div>
+      ) : listDenied ? (
+        // Checked BEFORE the empty state, and rendering only the i18n
+        // sentence: no status code, no server message, no row (#2532).
+        <div
+          className="px-4 py-6 text-sm text-muted-foreground flex items-center gap-2"
+          data-testid="record-attachments-denied"
+        >
+          <Lock className="h-4 w-4 shrink-0" />
+          {t('detail.attachmentsAccessDenied', {
+            defaultValue: "You don't have access to these attachments.",
+          })}
         </div>
       ) : rows.length === 0 ? (
         <div className="px-4 py-6 text-sm text-muted-foreground">

--- a/packages/app-shell/src/views/__tests__/RecordAttachmentsPanel.test.tsx
+++ b/packages/app-shell/src/views/__tests__/RecordAttachmentsPanel.test.tsx
@@ -94,6 +94,125 @@ describe('RecordAttachmentsPanel — server-denial error mapping (#2755)', () =>
   });
 });
 
+/**
+ * A denied list is not an empty list (#4269).
+ *
+ * A member denied the parent opened a record holding 2095+ attachments; the
+ * `sys_attachment` list read answered 403 and the panel rendered "No
+ * attachments yet. Upload a file to get started." — an assertion about the
+ * record's contents that the panel had no standing to make, plus an Upload
+ * the server would refuse.
+ *
+ * The 403 arrives as a THROW: the ObjectStack adapter's `find()` degrades only
+ * a non-authz 404 to `{ data: [], total: 0 }` and rethrows everything else, so
+ * the panel's `catch` is where the two verdicts were being merged.
+ */
+describe('RecordAttachmentsPanel — denied vs empty list (#4269)', () => {
+  /** The decorated shape the adapter rethrows: `httpStatus` + semantic code. */
+  function deniedListSource() {
+    return makeDataSource({
+      find: vi.fn(async () => {
+        throw Object.assign(
+          new Error(
+            "[Security] Access denied: operation 'find' on sys_attachment for user u-42 (2095 rows withheld)",
+          ),
+          { httpStatus: 403, code: 'PERMISSION_DENIED' },
+        );
+      }),
+    });
+  }
+
+  it('THE DEFECT — a 403 list read renders the denied state, never "No attachments yet"', async () => {
+    setup(deniedListSource());
+
+    await waitFor(() =>
+      expect(screen.getByTestId('record-attachments-denied')).toBeInTheDocument(),
+    );
+    expect(screen.getByText("You don't have access to these attachments.")).toBeInTheDocument();
+    // The empty state is reserved for a genuine 200-with-zero-rows.
+    expect(screen.queryByText(/No attachments yet/)).not.toBeInTheDocument();
+  });
+
+  it('withdraws the Upload affordance under a denied list', async () => {
+    setup(deniedListSource());
+
+    await waitFor(() =>
+      expect(screen.getByTestId('record-attachments-denied')).toBeInTheDocument(),
+    );
+    expect(screen.queryByRole('button', { name: 'Upload' })).not.toBeInTheDocument();
+  });
+
+  // The objectui#2532 failure mode — raw error dump, status code, leaked row
+  // counts — was ABSENT before this change and must stay absent: the denied
+  // state renders the i18n sentence and nothing sourced from the error.
+  it('leaks nothing from the error body — no status code, no server text, no row count', async () => {
+    setup(deniedListSource());
+
+    await waitFor(() =>
+      expect(screen.getByTestId('record-attachments-denied')).toBeInTheDocument(),
+    );
+    const panel = screen.getByTestId('record-attachments-panel');
+    expect(panel.textContent).not.toMatch(/403/);
+    expect(panel.textContent).not.toMatch(/2095/);
+    expect(panel.textContent).not.toMatch(/u-42/);
+    expect(panel.textContent).not.toMatch(/\[Security\]|Access denied: operation/);
+    // The denial is a state, not an error banner.
+    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
+  });
+
+  it('classifies a PERMISSION_DENIED code carrying no numeric status', async () => {
+    const dataSource = makeDataSource({
+      find: vi.fn(async () => {
+        const err: any = new Error('refused');
+        err.code = 'PERMISSION_DENIED';
+        throw err;
+      }),
+    });
+    setup(dataSource);
+
+    await waitFor(() =>
+      expect(screen.getByTestId('record-attachments-denied')).toBeInTheDocument(),
+    );
+    expect(screen.queryByRole('button', { name: 'Upload' })).not.toBeInTheDocument();
+  });
+
+  it('a genuine 200-with-zero-rows still renders the empty state AND the Upload button', async () => {
+    setup(makeDataSource({ find: vi.fn(async () => []) }));
+
+    await waitFor(() =>
+      expect(
+        screen.getByText('No attachments yet. Upload a file to get started.'),
+      ).toBeInTheDocument(),
+    );
+    expect(screen.getByRole('button', { name: 'Upload' })).toBeInTheDocument();
+    expect(screen.queryByTestId('record-attachments-denied')).not.toBeInTheDocument();
+  });
+
+  // PIN, not an endorsement. A non-authz failure keeps the pre-existing
+  // behaviour — swallowed to the empty state, Upload still offered. That is
+  // the same "unknown rendered as empty" defect one status over, but fixing it
+  // needs the loaded/unknown status vocabulary (#4235-style), not this
+  // denied-vs-empty split; filed separately. This test exists so the next
+  // change to that branch is a DELIBERATE one.
+  it('pins today’s behaviour for a NON-authz failure: still the empty state', async () => {
+    setup(
+      makeDataSource({
+        find: vi.fn(async () => {
+          throw new TypeError('Failed to fetch');
+        }),
+      }),
+    );
+
+    await waitFor(() =>
+      expect(
+        screen.getByText('No attachments yet. Upload a file to get started.'),
+      ).toBeInTheDocument(),
+    );
+    expect(screen.queryByTestId('record-attachments-denied')).not.toBeInTheDocument();
+    expect(screen.getByRole('button', { name: 'Upload' })).toBeInTheDocument();
+  });
+});
+
 describe('RecordAttachmentsPanel — authenticated signed-URL download (#2970)', () => {
   it('fetches /files/:id/url with auth and opens the signed URL', async () => {
     const openSpy = vi.spyOn(window, 'open').mockImplementation(() => null);

--- a/packages/i18n/src/locales/ar.ts
+++ b/packages/i18n/src/locales/ar.ts
@@ -900,6 +900,7 @@ const ar = {
     attachmentDownloadDenied: "ليس لديك صلاحية تنزيل هذا المرفق.",
     attachmentAuthRequired: "يرجى تسجيل الدخول لتنزيل هذا المرفق.",
     attachmentPermissionDenied: "ليس لديك إذن للقيام بذلك.",
+    attachmentsAccessDenied: "ليس لديك حق الوصول إلى هذه المرفقات.",
     unifiedDiff: "عرض موحد",
     sideBySideDiff: "عرض جنباً إلى جنب",
     noChanges: "لا توجد تغييرات",

--- a/packages/i18n/src/locales/de.ts
+++ b/packages/i18n/src/locales/de.ts
@@ -894,6 +894,7 @@ const de = {
     attachmentDownloadDenied: "Sie sind nicht berechtigt, diesen Anhang herunterzuladen.",
     attachmentAuthRequired: "Bitte melden Sie sich an, um diesen Anhang herunterzuladen.",
     attachmentPermissionDenied: "Sie haben keine Berechtigung für diese Aktion.",
+    attachmentsAccessDenied: "Sie haben keinen Zugriff auf diese Anhänge.",
     unifiedDiff: "Einheitliche Ansicht",
     sideBySideDiff: "Nebeneinander-Ansicht",
     noChanges: "Keine Änderungen",

--- a/packages/i18n/src/locales/en.ts
+++ b/packages/i18n/src/locales/en.ts
@@ -1030,6 +1030,7 @@ const en = {
     attachmentDownloadDenied: "You don't have access to download this attachment.",
     attachmentAuthRequired: 'Please sign in to download this attachment.',
     attachmentPermissionDenied: "You don't have permission to do that.",
+    attachmentsAccessDenied: "You don't have access to these attachments.",
     // Diff
     unifiedDiff: 'Unified diff',
     sideBySideDiff: 'Side-by-side diff',

--- a/packages/i18n/src/locales/es.ts
+++ b/packages/i18n/src/locales/es.ts
@@ -898,6 +898,7 @@ const es = {
     attachmentDownloadDenied: "No tiene acceso para descargar este adjunto.",
     attachmentAuthRequired: "Inicie sesión para descargar este adjunto.",
     attachmentPermissionDenied: "No tiene permiso para hacer eso.",
+    attachmentsAccessDenied: "No tiene acceso a estos adjuntos.",
     unifiedDiff: "Vista unificada",
     sideBySideDiff: "Vista lado a lado",
     noChanges: "Sin cambios",

--- a/packages/i18n/src/locales/fr.ts
+++ b/packages/i18n/src/locales/fr.ts
@@ -896,6 +896,7 @@ const fr = {
     attachmentDownloadDenied: "Vous n'avez pas accès pour télécharger cette pièce jointe.",
     attachmentAuthRequired: "Veuillez vous connecter pour télécharger cette pièce jointe.",
     attachmentPermissionDenied: "Vous n'avez pas la permission de faire cela.",
+    attachmentsAccessDenied: "Vous n'avez pas accès à ces pièces jointes.",
     unifiedDiff: "Vue unifiée",
     sideBySideDiff: "Vue côte à côte",
     noChanges: "Aucune modification",

--- a/packages/i18n/src/locales/ja.ts
+++ b/packages/i18n/src/locales/ja.ts
@@ -905,6 +905,7 @@ const ja = {
     attachmentDownloadDenied: "この添付ファイルをダウンロードする権限がありません。",
     attachmentAuthRequired: "この添付ファイルをダウンロードするにはサインインしてください。",
     attachmentPermissionDenied: "この操作を行う権限がありません。",
+    attachmentsAccessDenied: "これらの添付ファイルにアクセスする権限がありません。",
     unifiedDiff: "統合差分",
     sideBySideDiff: "横並び差分",
     noChanges: "変更なし",

--- a/packages/i18n/src/locales/ko.ts
+++ b/packages/i18n/src/locales/ko.ts
@@ -894,6 +894,7 @@ const ko = {
     attachmentDownloadDenied: "이 첨부파일을 다운로드할 권한이 없습니다.",
     attachmentAuthRequired: "이 첨부파일을 다운로드하려면 로그인하세요.",
     attachmentPermissionDenied: "이 작업을 수행할 권한이 없습니다.",
+    attachmentsAccessDenied: "이 첨부파일에 접근할 권한이 없습니다.",
     unifiedDiff: "통합 보기",
     sideBySideDiff: "나란히 보기",
     noChanges: "변경 없음",

--- a/packages/i18n/src/locales/pt.ts
+++ b/packages/i18n/src/locales/pt.ts
@@ -895,6 +895,7 @@ const pt = {
     attachmentDownloadDenied: "Você não tem acesso para baixar este anexo.",
     attachmentAuthRequired: "Faça login para baixar este anexo.",
     attachmentPermissionDenied: "Você não tem permissão para fazer isso.",
+    attachmentsAccessDenied: "Você não tem acesso a estes anexos.",
     unifiedDiff: "Visualização unificada",
     sideBySideDiff: "Visualização lado a lado",
     noChanges: "Sem alterações",

--- a/packages/i18n/src/locales/ru.ts
+++ b/packages/i18n/src/locales/ru.ts
@@ -913,6 +913,7 @@ const ru = {
     attachmentDownloadDenied: "У вас нет доступа для скачивания этого вложения.",
     attachmentAuthRequired: "Войдите в систему, чтобы скачать это вложение.",
     attachmentPermissionDenied: "У вас нет разрешения на это действие.",
+    attachmentsAccessDenied: "У вас нет доступа к этим вложениям.",
     unifiedDiff: "Единое представление",
     sideBySideDiff: "Параллельное представление",
     noChanges: "Нет изменений",

--- a/packages/i18n/src/locales/zh.ts
+++ b/packages/i18n/src/locales/zh.ts
@@ -974,6 +974,7 @@ const zh = {
     attachmentDownloadDenied: '您无权下载此附件。',
     attachmentAuthRequired: '请登录后再下载此附件。',
     attachmentPermissionDenied: '您没有执行此操作的权限。',
+    attachmentsAccessDenied: '您没有访问这些附件的权限。',
     // Diff
     unifiedDiff: '统一视图',
     sideBySideDiff: '并排视图',

